### DISCUSSION TITLE
Cache Replication with 2 replicas not working

### DISCUSSION THREAD

**Discussion by 98jan** [keycloak/keycloak#10125] [Archive]
Hello,

I started configuring a Keycloak.X server and so far mostly everything works fine.
Now I started to increase the replica in our Kubernetes Cluster from 1 to 2 and I am having some issues.

First of all the two Keycloak.X pods can find each other and successfully cluster, like the following message state:
`2022-02-10 14:19:38,547 INFO  [org.inf.CLUSTER] (keycloak-cache-init) ISPN000094: Received new cluster view for channel ISPN: [keycloak-6d47b457b4-wzwhk-44536|1] (2) [keycloak-6d47b457b4-wzwhk-44536, keycloak-6d47b457b4-4km5j-1715]`

As database I am using MySQL.

But if I open the URL in my browser the webpage looks broken, see following image.
Also login isn't working, looks like the caching or sth. is missing?
![image](https://user-images.githubusercontent.com/46779261/153431419-fc811331-8da4-4f0f-aac7-fc98c219fb14.png)
![image](https://user-images.githubusercontent.com/46779261/153432737-4f25c567-705c-47bd-939b-da5448d54124.png)

I am getting following warning, but don't know if this causes an issue:
`2022-02-10 14:19:33,984 WARN  [org.inf.CONFIG] (keycloak-cache-init) ISPN000569: Unable to persist Infinispan internal caches as no global state enabled`

I am using following arguments to start the keycloak.X:
`args: ["start","--auto-build","--http-port=8080","-Djgroups.dns.query=<dns-query>","--hostname=<domain>","--proxy=edge","--db=mysql","--db-url=jdbc:mysql://mysql:3306/<databasename>","--db-username=<username>","--db-password=<password>"]`

Could you please help me.

Best regards,
Jan

**Comment by 98jan** [comment]
Got it working after adding "--cache=ispn" found it in following documentation: https://www.keycloak.org/server/caching

**Comment by pedroigor** [comment]
@98jan To deploy in Kubernetes, you have a specific cache stack configuration. This configuration is based on the defaults from Infinispan for deploying a cluster on Kubernetes.

Could you please check this discussion https://github.com/keycloak/keycloak/issues/9644#issuecomment-1016850466 ?

**Reply by 98jan** [reply]
got it working there were some configuration parameters missing which I thought were default, after the release of documentation on Friday with full support of Quarkus, I found the needed configurations, thanks for your help.

**Reply by pedroigor** [reply]
Glad the documentation works for you. Thanks.

**Comment by geowatson** [comment]
Have the same problem but now in my k8s cluster the following message occurs on every instance of kc:
```
2022-03-10 07:55:12,061 INFO  [org.jgroups.protocols.pbcast.GMS] (keycloak-cache-init) keycloak-0-51857: no members discovered after 2002 ms: creating cluster as coordinator
```

**Reply by geowatson** [reply]
building with vars:
```
  - name: KC_CACHE
    value: ispn
  - name: KC_CACHE_STACK
    value: kubernetes
```

**Reply by 98jan** [reply]
could you check if following property helps you to let it find:

`args: ["start","--auto-build","-Djgroups.dns.query=servicename.namespace.cluster.local"]`

The "cluster.local" can differ to the used domain for internal resolving.

Also the --auto-build isn't needed for running in production.

Also check this, there I found everything I needed except kc_cache and stack option:
https://gist.github.com/pedroigor/e1476a41b544d15c1bd59155aad4f6ad

**Reply by pedroigor** [reply]
Thanks, @98jan.

@geowatson We have updated the documentation to mention this additional setting to configure the `DNS_PING` protocol.

**Reply by geowatson** [reply]
just checked, the problem is that my cluster uses NodePort Service (not headless), is this possible to run KC in NodePort k8s service?

**Reply by pedroigor** [reply]
TBH, I don't know. We have been using a headless service because of recommendations from more experienced people in our team. So DNS_PING can use this service to query the DNS and discover pods.
